core/sdk-core/src/main/java/software/amazon/awssdk/core/waiters/WaiterAcceptor.java

```diff
@@ -0,0 +1,214 @@
+     *
+     * @param errorPredicate the {@link Throwable} predicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> successOnExceptionAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.SUCCESS;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+
+    /**
+     * Creates an error waiter acceptor which determines if the exception should transition the waiter to failure state
+     *
+     * @param errorPredicate the {@link Throwable} predicate
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorOnExceptionAcceptor(Predicate<Throwable> errorPredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(Throwable t) {
+                return errorPredicate.test(t);
+            }
+        };
+    }
+
+    /**
+     * Creates a success waiter acceptor which determines if the exception should transition the waiter to success state
+     *
+     * @param responsePredicate the predicate of the response
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorOnResponseAcceptor(Predicate<T> responsePredicate) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(T response) {
+                return responsePredicate.test(response);
+            }
+        };
+    }
+
+    /**
+     * Creates a success waiter acceptor which determines if the exception should transition the waiter to success state
+     *
+     * @param responsePredicate the predicate of the response
+     * @param <T> the response type
+     * @return a {@link WaiterAcceptor}
+     */
+    static <T> WaiterAcceptor<T> errorOnResponseAcceptor(Predicate<T> responsePredicate, String message) {
+        return new WaiterAcceptor<T>() {
+            @Override
+            public WaiterState waiterState() {
+                return WaiterState.FAILURE;
+            }
+
+            @Override
+            public boolean matches(T response) {
+                return responsePredicate.test(response);
+            }
+
+            @Override
+            public Optional<String> message() {
+                return Optional.of(message);
+            }
+        };
+    }
```

Comment:
Validate the message is not-null so that it doesn't fail slow? Or use Optional.ofNullable.

Reply:
Ah, good catch!